Source/JavaScriptCore/yarr/YarrInterpreter.cpp

```diff
@@ -408,47 +490,76 @@ class Interpreter {
         return false;
     }
 
-    bool checkCharacter(int testChar, unsigned negativeInputOffset)
+    bool checkCharacter(ByteTerm& term, unsigned negativeInputOffset)
     {
-        return testChar == input.readChecked(negativeInputOffset);
+        ASSERT(term.isCharacterType());
+        if (term.matchDirection() == Forward)
+            return term.atom.patternCharacter == input.readChecked(negativeInputOffset);
+
+        if (!input.isValidOffset(negativeInputOffset))
+            return false;
+
+        return term.atom.patternCharacter == input.tryReadBackward(negativeInputOffset);
     }
 
-    bool checkSurrogatePair(int testUnicodeChar, unsigned negativeInputOffset)
+    bool checkSurrogatePair(ByteTerm& term, unsigned negativeInputOffset)
     {
-        return testUnicodeChar == input.readSurrogatePairChecked(negativeInputOffset);
+        ASSERT(term.isCharacterType());
+        if (term.matchDirection() == Forward)
+            return term.atom.patternCharacter == input.readSurrogatePairChecked(negativeInputOffset);
+
+        return term.atom.patternCharacter == input.readBackwardSurrogatePairChecked(negativeInputOffset);
+
     }
 
-    bool checkCasedCharacter(int loChar, int hiChar, unsigned negativeInputOffset)
+    bool checkCasedCharacter(ByteTerm& term, unsigned negativeInputOffset)
     {
-        int ch = input.readChecked(negativeInputOffset);
-        return (loChar == ch) || (hiChar == ch);
+        ASSERT(term.isCasedCharacterType());
+
+        int ch = term.matchDirection() == Forward ? input.readChecked(negativeInputOffset) : input.tryReadBackward(negativeInputOffset);
```

Comment:
Ditto, why isn't it `readBackwardChecked`?

Reply:
See above.